crates/red_knot_python_semantic/src/types/infer.rs

```
@@ -3653,16 +3654,44 @@ impl<'db> TypeInferenceBuilder<'db> {
             .member(db, &attr.id)
             .unwrap_with_diagnostic(|lookup_error| match lookup_error {
                 LookupError::Unbound => {
+                    let bound_on_instance = match value_type {
+                    Type::ClassLiteral(ClassLiteralType { class }) => {
+                        !class.instance_member(self.db(), attr).0.is_unbound()
+                    }
+                    Type::SubclassOf(subclass_of @ SubclassOfType { .. }) => {
+                        match subclass_of.subclass_of() {
+                            ClassBase::Class(class) => {
+                                !class.instance_member(self.db(), attr).0.is_unbound()
+                            }
+                            ClassBase::Dynamic(_) => false,
```

I think we could even be a little stricter here and mark this branch `unreachable!`? Because `type[Any]` has all attributes, so we can never get `Unbound` back from getting a member off it.

The main advantage would be that in some sense neither `true` nor `false` seems right here, so it might make the code a bit clearer, and it would mean we fail fast if we violate this invariant. The disadvantage would be that if we somehow ship a very subtle violation of the invariant, it could cause a panic for a user.

I don't feel strongly about using `unreachable!`, open to whatever you prefer. But if we don't use `unreachable!`, I feel like `true` makes slightly more sense here than `false`, since `Any` also has all members.

---

checked the codebase, seems like unreachable is being used in a ocuple of other places and also verified this invariant (mostly for myself). I think it is reasonable to use unreachable here. Some tests will fail quickly if it stops holding true.